cmd/thanos/sidecar.go

```diff
@@ -214,6 +239,33 @@ func runSidecar(
 		g.Add(func() error {
 			defer gcsClient.Close()
 
+			return runutil.Repeat(30*time.Second, ctx.Done(), func() error {
+				s.Sync(ctx)
+
+				minTime, _, err := s.Timestamps()
+				if err != nil {
+					level.Warn(logger).Log("msg", "reading timestamps failed", "err", err)
+				} else {
+					peer.SetTimestamps(minTime, math.MaxInt64)
+				}
+				return nil
+			})
+		}, func(error) {
+			cancel()
+		})
+	} else if s3Config.Validate() != false {
+		var bkt objstore.Bucket
+		bkt, err = s3.NewBucket(s3Config, reg)
+		if err != nil {
+			return errors.Wrap(err, "create s3 client")
+		}
+		bkt = objstore.BucketWithMetrics(*s3Config.Bucket, bkt, reg)
+
+		s := shipper.New(logger, nil, dataDir, bkt, externalLabels.Get)
```

Comment:
Can we reuse the same shipper and below group.Add and just switch between different objstore implementations?

Reply:
Ah ok, see your point regarding `.Close()`. Please see my suggestion with `closeFn` method